spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/core/subscriber/PubSubSubscriberTemplate.java

```diff
@@ -261,6 +269,11 @@ public SubscriberFactory getSubscriberFactory() {
 						modifyAckDeadline(subscriptionName, ackIds, ackDeadlineSeconds));
 	}
 
+	@Override
+	public void destroy() throws Exception {
+		this.ackExecutor.shutdown();
```

Comment:
See my comment in the `setAckExecutor()` and you definitely can't call `shutdown()` on the external executor. That's just not your responsibility to control lifecycle of the injected objects. However with the `!this.ackExecutorSetExplicitly` you can destroy that default internal `Executors.newSingleThreadExecutor()`. Only the point that you have to cast since we would like to expect from end-users a generic `Executor`.

Reply:
I am looking at the `RedisLockRegistry` for inspiration, since it uses this pattern. It seems to me that when creating an executor as part of field initialization, there is a possibility of it never being shutdown [when a custom `Executor` is injected](https://github.com/spring-projects/spring-integration/blob/88c7646ed9eaea2fe755485e1676ad130eb391a9/spring-integration-redis/src/main/java/org/springframework/integration/redis/util/RedisLockRegistry.java#L151), although the object itself will be garbage collected.

I could shut down the pool in the custom executor setter, although that's not ideal since clients would expect setters to not do extra work:

````
	public void setAckExecutor(Executor ackExecutor) {
		Assert.notNull(ackExecutor, "ackExecutor can't be null.");
                ((ExecutorService)this.ackExecutor).shutdown();
		this.ackExecutor = ackExecutor;
		this.ackExecutorSetExplicitly = true;
	}
````

Another option is to remove field initialization, and check instead for whether the field is null in `doBatchedAsyncOperation()`:

````
 doBatchedAsyncOperation(...) {
  if (this.ackExecutor == null) {
    this.ackExecutor = Executors.newCachedThreadPool();
  }
  // use the executor here
}
````

A third possibility is to have a `defaultAckExecutor` object that always gets destroyed, and a `customAckExecutor` that can be user-injected. Then in destroy(), it is safe to always shut down `defaultAckExecutor`. That feels like the cleanest approach to me.

What do you think?

Reply:
I understand what you mean -- that in real world, it won't cause problems, even though it's in the ExecutorService contract that unused objects should be shut down.

On the other hand... we don't typically prevent users from setting template properties after some work has already been done, so getting a new executor after the single-threaded executor started work is weird, but not impossible.

I'll go with `defaultAckExecutor`.